Source/CesiumRuntime/Private/CesiumGltfPrimitiveComponent.h

```diff
@@ -22,14 +28,9 @@ struct Model;
 struct MeshPrimitive;
 } // namespace CesiumGltf
 
-UCLASS()
-class UCesiumGltfPrimitiveComponent : public UStaticMeshComponent {
-  GENERATED_BODY()
-
+class CesiumGltfPrimitiveBase {
```

Comment:
Yeah sorry Tim I'm not a huge fan of this multiple inheritance approach. My main complaints are:

1. It requires some implementation to move to the .h file, and the corresponding headers to be included. I'm pretty big on keeping as much as we can out of the headers because it increases compile time and object file size (we have to ship object files for every platform, so it adds up).
2. The `_uobject` bit. It's hacky. Especially since it's public.
3. Multiple implementation inheritance is widely considered something to avoid.

I mean, these things aren't huge, so if we were getting some significant advantage from it, I'd be willing to let it go. But I don't think we are. The `UpdateTransformFromCesium` can easily be handled by delegating to a helper function in the two classes. It probably doesn't even need to be a template; I think everything that method uses is available on `UPrimitiveComponent` which is the common base class.

As for all those public fields in `CesiumGltfPrimitiveBase`... well, they're a bit of a mess anyway, aren't they? Ideally I'd like to see thoughtful, Blueprint-accessible accessors for the ones that make sense, and the rest made private. But that's a lot of scope creep for this PR, so I'd be happy enough if you just copied them into the `UCesiumGltfInstancedComponent` (a template function in CesiumGltfCompnent.cpp can be used to set them). Or, better yet, moved them to a common object that is aggregated in both classes (this is a breaking change, but no one is likely to care. We just need to document it). Or... better still... do we even need most of them? A lot of them are probably only relevant for things that aren't supported yet, like metadata. So just leaving them out of the new class for now is an excellent option.

Reply:
> Yeah sorry Tim I'm not a huge fan of this multiple inheritance approach. My main complaints are:
> 
> 1. It requires some implementation to move to the .h file, and the corresponding headers to be included. I'm pretty big on keeping as much as we can out of the headers because it increases compile time and object file size (we have to ship object files for every platform, so it adds up).
> 2. The `_uobject` bit. It's hacky. Especially since it's public.
> 3. Multiple implementation inheritance is widely considered something to avoid.

While I don't agree with your points in general, I have to agree in the context of an Unreal plugin. _uobject is only needed because regular dynamic_cast isn't available; Unreal turns off rtti. Otherwise, I think multiple inheritance is a fine way to add a common facet of data and behavior to an inheritance chain that has a very obvious "single inheritance" axis. Again, downcasting is very awkward in the Unreal universe.

I've committed the change. I think the the results a messier and more verbose than using multiple inheritance, but then there's no accounting for taste 😄 
> 
> 
> As for all those public fields in `CesiumGltfPrimitiveBase`... well, they're a bit of a mess anyway, aren't they?

Not really for me to say.

 >Ideally I'd like to see thoughtful, Blueprint-accessible accessors for the ones that make sense, and the rest made private. But that's a lot of scope creep for this PR, so I'd be happy enough if you just copied them into the `UCesiumGltfInstancedComponent` (a template function in CesiumGltfCompnent.cpp can be used to set them). Or, better yet, moved them to a common object that is aggregated in both classes (this is a breaking change, but no one is likely to care. We just need to document it). Or... better still... do we even need most of them? A lot of them are probably only relevant for things that aren't supported yet, like metadata. So just leaving them out of the new class for now is an excellent option.

I'm hoping to use them soon for instances. Maybe we can have another go-around then.

Reply:
> _uobject is only needed because regular dynamic_cast isn't available; Unreal turns off rtti.

I haven't looked at your changes yet, but a quick note here... Unreal has its own RTTI system for UObjects. You can downcast using the `Cast<T>` function. It works much like `dynamic_cast<T>`, e.g.: 

```cpp
UObject* pWhatever = ...;
UCesiumGltfPrimitiveComponent* pPrimitive = Cast<UCesiumGltfPrimitiveComponent>(pWhatever);
if (pPrimitive) {
  // ...
}
```